### ISSUE TITLE
scal_kernel in addmm bprop

### ISSUE THREAD

**Issue by adityaiitb** [NVIDIA/PyProf#162]
> hi Karthikeyan,
> 
> A few observations and remarks.
> 
> a) Are you using the branch `main`? That is the default branch. Can you switch to that branch and reinstall.
> b) In your earlier message the error occurred after around 6200 kernels. However, the `net.dict` file you shared has only 73 kernels.
> c) I ran the `net.dict` and ran into a different error. The error happens on line 48 due to `torch.zeros_like` which is confusing. According to the [docs](https://pytorch.org/docs/stable/generated/torch.zeros_like.html), the first input to `torch.zeros_like` should be a tensor. However, the first argument appears to be an int with value 0.
> 
> Can you please double check @karthi0804

Thanks for ur time @adityaiitb  
1.  I am using the main branch.
2. Sorry my mistake. I didn't copy the dict file fully into gist. Please have a look at the [new file](https://drive.google.com/file/d/1C3zN-I-MYmLMHM3Qjze-eALgPzrPHhFM/view?usp=sharing) now. 
Pls check line `11086` in net.dict file where I get the assertion error.

_Originally posted by @karthi0804 in https://github.com/NVIDIA/PyProf/issues/46#issuecomment-818346688_

**Comment by adityaiitb** [comment]
1. The issue is with the kernel `scal_kernel` in the bprop of `torch.addmm`. AFAI can tell, its not related to either the data gradient (`dX`) or the weight gradient (`dW`) or the bias gradient (`dB`). Its something else, which I have not observed before. The corresponding fprop call is on line 2494 of net.dict which comes from `intentionprediction/model.py:20`. What you can do temporarily, while I investigate, is to disable the assert `pyprof/prof/linear.py:115` and reinstall.

Given that you know where the error is coming from, if possible can you create a 10 line bug repro.

2. After you disable that assert and reinstall, you will hit another error at line 12391. This is confusing and is related to `torch.zeros_like` which I mentioned earlier. The first argument is an `int` instead of a `tensor`. From the file I can tell its coming from `torchmetrics/utilities/distributed.py:109`. Can you update this library and see if this error goes away? Again if you just want to get past the error temporarily for whatever reason you can comment out `prof/memory.py:44` and put something like `self.input = Tensor([10,10], "int")` and reinstall.

I understand this call is coming from a library, but if possible can you create a small repro.

As a side note, I noticed that you are using fp32 as your datatype which means you are not using any of the tensor core on Volta / Turing / Ampere. The training will go much faster if you use fp16 / mixed precision.

**Comment by adityaiitb** [comment]
I was able to create a repro for (2). The error happens for tensors which are scalars. I will fix it soon. Thanks for reporting.

```
import torch
import pyprof
pyprof.init()
a = torch.tensor(5).cuda()
b = torch.zeros_like(a)
```

**Comment by karthi0804** [comment]
@adityaiitb Thanks for the update. So you already created repro. I don't have to check anything. right? I will wait for the fix. 
I badly wanted to use fp16. Since I am using the graph neural network model, torch-geometric doesn't support fp16 completely (according to [this](https://github.com/rusty1s/pytorch_scatter/issues/191)) Pls let me know if I can speed up the training by any other means.

**Comment by adityaiitb** [comment]
If you can provide a repro for (1) that would be great. Not sure about torch-geometric. PyTorch now has `autocast`. You might want to try it out. If you do end up using `fp16` I would suggest making your tensor dimensions good multiples of 8.

**Comment by karthi0804** [comment]
Ok, I will try to provide. Another thing I noticed is any normal python operations using '+=' within model code throws an error. But I replaced that using `x = x + y` to avoid that errors. So I think there is no support for python __iadd__.
